drivers/i2c/i2c_esp32.c

```diff
@@ -0,0 +1,713 @@
+	I2C_ESP32_OP_STOP,
+	I2C_ESP32_OP_END
+};
+
+struct i2c_esp32_cmd {
+	u32_t num_bytes : 8;
+	u32_t ack_en : 1;
+	u32_t ack_exp : 1;
+	u32_t ack_val : 1;
+	u32_t opcode : 3;
+	u32_t : 17;
+	u32_t done : 1;
+};
+
+struct i2c_esp32_data {
+	u32_t dev_config;
+	u16_t address;
+
+	struct k_sem fifo_sem;
+	struct k_sem transfer_sem;
+};
+
+typedef void (*irq_connect_cb)(void);
+
+struct i2c_esp32_config {
+	int index;
+
+	irq_connect_cb connect_irq;
+
+	const struct {
+		int sda_out;
+		int sda_in;
+		int scl_out;
+		int scl_in;
+	} sig;
+
+	const struct {
+		int scl;
+		int sda;
+	} pins;
+
+	const struct {
+		int clk;
+		int rst;
+	} enable_bits;
+
+	const struct {
+		bool tx_lsb_first;
+		bool rx_lsb_first;
+	} mode;
+
+	const struct {
+		int source;
+		int line;
+	} irq;
+
+	const u32_t default_config;
+};
+
+static inline void set_mask32(u32_t v, u32_t mem_addr)
+{
+	sys_write32(sys_read32(mem_addr) | v, mem_addr);
+}
+
+static inline void clear_mask32(u32_t v, u32_t mem_addr)
+{
+	sys_write32(sys_read32(mem_addr) & ~v, mem_addr);
+}
+
+static void i2c_esp32_enable_peripheral(const struct i2c_esp32_config *config)
+{
+	set_mask32(config->enable_bits.clk, DPORT_PERIP_CLK_EN_REG);
+	clear_mask32(config->enable_bits.rst, DPORT_PERIP_RST_EN_REG);
+}
+
+static const char *i2c_esp32_get_gpio_for_pin(int pin)
+{
+	if (pin < 32) {
+#if defined(CONFIG_GPIO_ESP32_0)
+		return CONFIG_GPIO_ESP32_0_NAME;
+#else
+		return NULL;
+#endif /* CONFIG_GPIO_ESP32_0 */
+	}
+
+#if defined(CONFIG_GPIO_ESP32_1)
+	return CONFIG_GPIO_ESP32_1_NAME;
+#else
+	return NULL;
+#endif /* CONFIG_GPIO_ESP32_1 */
+}
+
+static int i2c_esp32_configure_pins(int pin, int matrix_out, int matrix_in)
+{
+	const int pin_mode = GPIO_DIR_OUT |
+			     GPIO_DS_DISCONNECT_LOW |
+			     GPIO_PUD_PULL_UP;
+	const char *device_name = i2c_esp32_get_gpio_for_pin(pin);
+	struct device *gpio;
+	int ret;
+
+	if (!device_name) {
+		return -EINVAL;
+	}
+	gpio = device_get_binding(device_name);
+	if (!gpio) {
+		return -EINVAL;
+	}
+
+	ret = gpio_pin_configure(gpio, pin, pin_mode);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = gpio_pin_write(gpio, pin, 1);
+	if (ret < 0) {
+		return ret;
+	}
+
+	esp32_rom_gpio_matrix_out(pin, matrix_out, false, false);
+	esp32_rom_gpio_matrix_in(pin, matrix_in, false);
+
+	return 0;
+}
+
+static int i2c_esp32_configure_frequency(const struct i2c_esp32_config *config,
+					 u32_t speed)
+{
+	static const u32_t frequencies[] = {
+		[I2C_SPEED_STANDARD] = KHZ(100),
+		[I2C_SPEED_FAST] = KHZ(400),
+		[I2C_SPEED_FAST_PLUS] = MHZ(1),
+		[I2C_SPEED_HIGH] = 0,
+		[I2C_SPEED_ULTRA] = 0
+	};
+	u32_t freq_hz = frequencies[speed];
+	u32_t period;
+
+	if (!freq_hz) {
+		return -ENOTSUP;
+	}
+
+	period = (APB_CLK_FREQ / freq_hz) / 2;
+
+	set_mask32(period << I2C_SCL_LOW_PERIOD_S,
+		   I2C_SCL_LOW_PERIOD_REG(config->index));
+	set_mask32(period << I2C_SCL_HIGH_PERIOD_S,
+		   I2C_SCL_HIGH_PERIOD_REG(config->index));
+
+	period /= 2; /* Set hold and setup times to 1/2th of period */
```

Comment:
Do these particular settings come from a design guide or something somewhere?  Might be worth citing it so people understand where the magic comes from.